### Exercises

#### Exercise 1

Read in the simulated gene expression dataset found in `gene_data.csv`. Then fit a multiple linear regression model and compute the RMSE for this model.

In [3]:
# YOUR CODE HERE FOR READING DATA AND FITTING REGRESSION MODEL
import pandas as pd
df = pd.read_csv("../data/gene_data.csv")
from sklearn.linear_model import LinearRegression
from sklearn.metrics import root_mean_squared_error

y = df["Trait"]
X = df.drop(columns="Trait")

full_model = LinearRegression()
full_model.fit(X, y)
full_preds = full_model.predict(X)

print("RMSE: ", root_mean_squared_error(y, full_preds))
df.head()

RMSE:  5.7604843529814405


,TP53,BRCA1,BRCA2,EGFR,MYC,KRAS,PTEN,APOE,TNF,VEGFA,...,PECAM1,VCAM1,IFNG,IL10,TGFB1,NOS3,PPARG,SREBF1,FASN,Trait
0,8.545177,1.930207,4.378380,4.026826,24.580835,4.443665,10.133618,44.363948,10.012965,13.597512,...,10.253655,8.300915,4.494035,5.102598,8.070732,2.742337,8.269346,9.127478,12.664307,397.425898
1,7.268998,7.739082,4.260046,187.088252,2.428403,5.527279,4.123282,6.489095,6.208605,4.678597,...,1.636358,3.702188,0.490370,3.770221,5.056198,2.061478,6.180184,44.558453,6.914127,423.246711
2,7.209494,10.807883,3.718422,7.500737,35.877357,5.237345,15.770932,9.650083,11.374552,7.330080,...,12.560870,3.541525,10.302735,18.979333,8.390883,1.193579,10.834315,4.480203,9.926243,366.093684
3,16.205869,12.612047,3.420913,66.814748,10.542371,5.922275,20.534795,217.900224,3.087735,1.801208,...,0.895400,15.023451,5.971124,6.261811,10.234016,72.412861,7.233361,18.423569,2.219636,407.067138
4,6.491159,4.214773,5.366326,1.062904,77.480568,6.173582,9.439733,13.281299,21.362738,3.986367,...,2.490854,15.439841,5.188251,7.293647,12.292569,13.123096,8.388838,4.485446,10.787298,343.912618


#### Exercise 2

##### Part 1

Split the data into training and test datasets using 20% of the data for testing and 80% for training.

In [4]:
# YOUR CODE FOR SPLITTING DATA INTO TRAINING AND TEST SETS
from sklearn.model_selection import train_test_split

X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.2)

##### Part 2

Fit the multiple linear regression model to the training data and then compute the test RMSE for the test dataset.

In [5]:
# YOUR CODE FOR FITTING MODEL ON TRAINING DATA AND COMPUTING TEST RMSE
train_model = LinearRegression()

train_model.fit(X_train, y_train)

test_preds = train_model.predict(X_test)

print("RMSE: ", root_mean_squared_error(y_test, test_preds))

RMSE:  56.34099249925032


##### Part 3

What do you observed about the train RMSE vs the test RMSE? Why do you think this is happening?

YOUR ANSWER HERE
The train RMSE 5.76 is lower than the test RMSE 56.34. This happens because the model starts fitting random variation instead of meaningful signal.

#### Exercise 3

##### Part 1

Use the `StandardScaler` class to compute means and standard deviations for each predictor in the training dataset.

In [9]:
# YOUR CODE HERE FOR CREATING THE STANDARDSCALER OBJECT FOR STANDARDIZATION
from sklearn.preprocessing import StandardScaler

scaler = StandardScaler()
scaler.fit(X_train)

print("Means:", scaler.mean_)
print("Std dev:", scaler.scale_)

Means: [ 8.33511222 14.98342245  9.4753804  19.45391636 16.14913467  7.58301122
 12.50328426 26.07948681 12.21036612  9.80485353 14.99420689 10.51604107
 15.2406221  11.34363229  8.11515883 22.29380434  7.43449201  7.41851596
  9.31499481 22.369782   19.27187866 11.74125066 13.56336533 37.74808069
 11.06051396 15.93946479 10.28050954 11.99733502  9.00390325  7.39406043
 18.69608819 25.98853829 12.63891933 13.77284349  7.69635378  9.13758004
 11.90029742  8.05663806  8.26419287  8.08361529  8.05087381  8.79227831
 10.1118637   9.23903117  8.52294427  7.95837082  8.75509259 12.12891891
  8.28158595 18.00885582  7.3579137  11.43410859 13.473907    7.35551487
 10.99242073  8.01980668  7.6393882  14.41068954 19.74683096  9.56420533
 11.2368903   7.7910701   9.78613817  8.02975943 30.74704805 10.71435915
 20.64684421 17.75025303 19.44012071  9.46349558 18.82561883 13.97807109
 14.1002834   7.40724034  9.11554977  8.72730086  9.31302447 10.24616373
  8.43125101  8.15704683  8.04959254 14.4392

##### Part 2

Standardize the predictors in the test dataset using the means and standard deviations computed in Part 1 with the `StandardScaler` class.

In [10]:
# YOUR CODE HERE FOR STANDARDIZING THE TEST DATA
X_test_scaled = scaler.transform(X_test)


#### Exercise 4

##### Part 1

Utilize the `LassoCV` class to use cross-validation to find the best value for $\alpha$ in the LASSO regression model and then compute the RMSE for the selected model.

In [11]:
# YOUR CODE HERE FOR PERFORMING CROSS VALIDATION AND COMPUTING RMSE FOR THE CROSS-VALIDATED MODEL
from sklearn.linear_model import LassoCV

lasso_CV_model = LassoCV(cv=10)
lasso_CV_model.fit(X_train_scaled, y_train)

lasso_CV_preds = lasso_CV_model.predict(X_test_scaled)

print("Selected alpha: ", lasso_CV_model.alpha_)
print("RMSE: ", root_mean_squared_error(y_test, lasso_CV_preds))

Selected alpha:  0.7597787682741194
RMSE:  21.317709510412822


##### Part 2

In your own words, explain what `LassoCV` is doing. Then explain the difference between train, validation, and test datasets.

YOUR ANSWER HERE
LassoCV is a method that uses Lasso regression to find the best alpha value by trying different values and selecting the one that gives the lowest cross-validation error.
The training dataset determines how the response variable depends on the predictor variables by learning the model parameters.
The validation dataset tries to get the best value of alpha after compared different alpha to optimize the model.
The test dataset provides the final values of the model’s predictive process and evaluates how the final model performs on unseen data.

##### Part 3

What is overfitting? How does LASSO regression utilize regularization to prevent overfitting?

YOUR ANSWER HERE
Overfitting is the result of a model that performs very well on training data but not well on new observations.
LASSO regression utilizes regularization by using a penalty that reduces model complexity to prevent overfitting.

##### Part 4

What exactly is the LASSO objective function designed to do and how does it prevent overfitting?

YOUR ANSWER HERE
The LASSO objective function is designed to reduce prediction error and ensure the model is both accurate and simple.
It prevents overfitting by using small coefficients to reduce complexity.

##### Part 5

Extract the coefficient estimates from the fitted model to determine which of the predictors are "active" (non-zero).

In [14]:
# YOUR CODE FOR EXTRACTING COEFFICIENTS HERE
active_predictors = {
    X.columns[i]: lasso_CV_model.coef_[i]
    for i in range(len(X.columns))
    if lasso_CV_model.coef_[i] != 0
}

print(active_predictors)

{'EGFR': 2.101916141991862, 'MYC': -0.2090878776681813, 'APOE': 2.4806174689606824, 'VEGFA': -0.2763171415037262, 'IL6': 1.519214708839394, 'MTHFR': 5.023877384149876, 'BRAF': 0.3764654215854834, 'ERBB2': -1.5390811708987686, 'SMAD4': 3.6531331926872785, 'KIT': -0.44506116214160657, 'PDGFRA': 1.7949880214715825, 'RB1': -7.640895845361323, 'NOTCH1': 0.9868265031429434, 'CCND1': -1.8652935918840161, 'CDH1': -16.35245941506166, 'VHL': 4.389276694199673, 'FLT3': 2.1669278025060708, 'ATM': 0.002256212927447336, 'PALB2': -0.076523508954661, 'AR': -0.7347235112941306, 'SMARCA4': -0.39809237605963027, 'TGFBR2': 1.2147442580635195, 'ROS1': -3.076847303352653, 'CDK6': 0.8864345654498195, 'RUNX1': 0.8703649358057183, 'DNMT3A': 7.027140853721261, 'TET2': -1.2722844982040484, 'GNAS': -2.180103429887883, 'HRAS': 1.6818805158025318, 'KLF4': 0.9413578986840727, 'CXCR4': 1.449465148994225, 'MMP2': -1.1837009156217553, 'COL1A1': 0.3244934034579264, 'COL3A1': -13.792025247824256, 'ITGB1': -0.499382528386